8263087: Add a MethodHandle combinator that switches over a set of MethodHandles

JornVernee opened the pull request:
This patch adds a `tableSwitch` combinator that can be used to switch over a set of method handles given an index, with a fallback in case the index is out of bounds, much like the `tableswitch` bytecode. Here is a description of how it works (copied from the javadoc):

     Creates a table switch method handle, which can be used to switch over a set of target
     method handles, based on a given target index, called selector.

     For a selector value of {@code n}, where {@code n} falls in the range {@code [0, N)},
     and where {@code N} is the number of target method handles, the table switch method
     handle will invoke the n-th target method handle from the list of target method handles.

     For a selector value that does not fall in the range {@code [0, N)}, the table switch
     method handle will invoke the given fallback method handle.

     All method handles passed to this method must have the same type, with the additional
     requirement that the leading parameter be of type {@code int}. The leading parameter
     represents the selector.

     Any trailing parameters present in the type will appear on the returned table switch
     method handle as well. Any arguments assigned to these parameters will be forwarded,
     together with the selector value, to the selected method handle when invoking it.

The combinator does not support specifying the starting index, so the switch cases always run from 0 to however many target handles are specified. A starting index can be added manually with another combination step that filters the input index by adding or subtracting a constant from it, which does not affect performance. One of the reasons for not supporting a starting index is that it allows for more lambda form sharing, but also simplifies the implementation somewhat. I guess an open question is if a convenience overload should be added for that case?

Lookup switch can also be simulated by filtering the input through an injection function that translates it into a case index, which has also proven to have the ability to have comparable performance to, or even better performance than, a bytecode-native `lookupswitch` instruction. I plan to add such an injection function to the runtime libraries in the future as well. Maybe at that point it could be evaluated if it's worth it to add a lookup switch combinator as well, but I don't see an immediate need to include it in this patch.

The current bytecode intrinsification generates a call for each switch case, which guarantees full inlining of the target method handles. Alternatively we could only have 1 callsite at the end of the switch, where each case just loads the target method handle, but currently this does not allow for inlining of the handles, since they are not constant.

Maybe a future C2 optimization could look at the receiver input for invokeBasic call sites, and if the input is a phi node, clone the call for each constant input of the phi. I believe that would allow simplifying the bytecode without giving up on inlining.

Some numbers from the added benchmarks:
```
Benchmark                                        (numCases)  (offset)  (sorted)  Mode  Cnt   Score   Error  Units
MethodHandlesTableSwitchConstant.testSwitch               5         0       N/A  avgt   30   4.186 � 0.054  ms/op
MethodHandlesTableSwitchConstant.testSwitch               5       150       N/A  avgt   30   4.164 � 0.057  ms/op
MethodHandlesTableSwitchConstant.testSwitch              10         0       N/A  avgt   30   4.124 � 0.023  ms/op
MethodHandlesTableSwitchConstant.testSwitch              10       150       N/A  avgt   30   4.126 � 0.025  ms/op
MethodHandlesTableSwitchConstant.testSwitch              25         0       N/A  avgt   30   4.137 � 0.042  ms/op
MethodHandlesTableSwitchConstant.testSwitch              25       150       N/A  avgt   30   4.113 � 0.016  ms/op
MethodHandlesTableSwitchConstant.testSwitch              50         0       N/A  avgt   30   4.118 � 0.028  ms/op
MethodHandlesTableSwitchConstant.testSwitch              50       150       N/A  avgt   30   4.127 � 0.019  ms/op
MethodHandlesTableSwitchConstant.testSwitch             100         0       N/A  avgt   30   4.116 � 0.013  ms/op
MethodHandlesTableSwitchConstant.testSwitch             100       150       N/A  avgt   30   4.121 � 0.020  ms/op
MethodHandlesTableSwitchOpaqueSingle.testSwitch           5         0       N/A  avgt   30   4.113 � 0.009  ms/op
MethodHandlesTableSwitchOpaqueSingle.testSwitch           5       150       N/A  avgt   30   4.149 � 0.041  ms/op
MethodHandlesTableSwitchOpaqueSingle.testSwitch          10         0       N/A  avgt   30   4.121 � 0.026  ms/op
MethodHandlesTableSwitchOpaqueSingle.testSwitch          10       150       N/A  avgt   30   4.113 � 0.021  ms/op
MethodHandlesTableSwitchOpaqueSingle.testSwitch          25         0       N/A  avgt   30   4.129 � 0.028  ms/op
MethodHandlesTableSwitchOpaqueSingle.testSwitch          25       150       N/A  avgt   30   4.105 � 0.019  ms/op
MethodHandlesTableSwitchOpaqueSingle.testSwitch          50         0       N/A  avgt   30   4.097 � 0.021  ms/op
MethodHandlesTableSwitchOpaqueSingle.testSwitch          50       150       N/A  avgt   30   4.131 � 0.037  ms/op
MethodHandlesTableSwitchOpaqueSingle.testSwitch         100         0       N/A  avgt   30   4.135 � 0.025  ms/op
MethodHandlesTableSwitchOpaqueSingle.testSwitch         100       150       N/A  avgt   30   4.139 � 0.145  ms/op
MethodHandlesTableSwitchRandom.testSwitch                 5         0      true  avgt   30   4.894 � 0.028  ms/op
MethodHandlesTableSwitchRandom.testSwitch                 5         0     false  avgt   30  11.526 � 0.194  ms/op
MethodHandlesTableSwitchRandom.testSwitch                 5       150      true  avgt   30   4.882 � 0.025  ms/op
MethodHandlesTableSwitchRandom.testSwitch                 5       150     false  avgt   30  11.532 � 0.034  ms/op
MethodHandlesTableSwitchRandom.testSwitch                10         0      true  avgt   30   5.065 � 0.076  ms/op
MethodHandlesTableSwitchRandom.testSwitch                10         0     false  avgt   30  13.016 � 0.020  ms/op
MethodHandlesTableSwitchRandom.testSwitch                10       150      true  avgt   30   5.103 � 0.051  ms/op
MethodHandlesTableSwitchRandom.testSwitch                10       150     false  avgt   30  12.984 � 0.102  ms/op
MethodHandlesTableSwitchRandom.testSwitch                25         0      true  avgt   30   8.441 � 0.165  ms/op
MethodHandlesTableSwitchRandom.testSwitch                25         0     false  avgt   30  13.371 � 0.060  ms/op
MethodHandlesTableSwitchRandom.testSwitch                25       150      true  avgt   30   8.628 � 0.032  ms/op
MethodHandlesTableSwitchRandom.testSwitch                25       150     false  avgt   30  13.542 � 0.020  ms/op
MethodHandlesTableSwitchRandom.testSwitch                50         0      true  avgt   30   4.701 � 0.015  ms/op
MethodHandlesTableSwitchRandom.testSwitch                50         0     false  avgt   30  13.562 � 0.063  ms/op
MethodHandlesTableSwitchRandom.testSwitch                50       150      true  avgt   30   7.991 � 3.111  ms/op
MethodHandlesTableSwitchRandom.testSwitch                50       150     false  avgt   30  13.543 � 0.088  ms/op
MethodHandlesTableSwitchRandom.testSwitch               100         0      true  avgt   30   4.712 � 0.020  ms/op
MethodHandlesTableSwitchRandom.testSwitch               100         0     false  avgt   30  13.600 � 0.085  ms/op
MethodHandlesTableSwitchRandom.testSwitch               100       150      true  avgt   30   4.676 � 0.011  ms/op
MethodHandlesTableSwitchRandom.testSwitch               100       150     false  avgt   30  13.476 � 0.043  ms/op
```

Testing:
- [x] Running of included benchmarks
- [x] Inspecting inlining trace and verifying method handle targets are inlined
- [x] Running TestTableSwitch test (currently the only user of the new code)
- [x] Running java/lang/invoke tests (just in case)
- [x] Some manual testing

Thanks,
Jorn

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8263087](https://bugs.openjdk.java.net/browse/JDK-8263087): Add a MethodHandle combinator that switches over a set of MethodHandles


### Reviewers
 * [Claes Redestad](https://openjdk.java.net/census#redestad) (@cl4es - **Reviewer**) ⚠️ Review applies to 80a706f1db7aef09830ae82b8bea0652c940fc55


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3401/head:pull/3401` \
`$ git checkout pull/3401`

Update a local copy of the PR: \
`$ git checkout pull/3401` \
`$ git pull https://git.openjdk.java.net/jdk pull/3401/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3401`

View PR using the GUI difftool: \
`$ git pr show -t 3401`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3401.diff">https://git.openjdk.java.net/jdk/pull/3401.diff</a>

</details>

JornVernee commented:
/csr

JornVernee commented:
> yes, for all the switches, pattern-switch, enum-switch but not for the string switch which requires a lookup switch.
Can you outline how to use the tableswitch combinator in the case of a switch on strings ?

Jan Lahoda has made several good examples of that: https://github.com/lahodaj/jdk/blob/switch-bootstrap/src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java

I ran some benchmarks:

![results](http://cr.openjdk.java.net/~jvernee/StringSwitch_10.svg)

Here, 'legacy' is what C2 does.

JornVernee commented:
I've uploaded a benchmark that simulates a lookup switch using the tableSwitch combinator as well, using a HashMap lookup as a filter: https://github.com/openjdk/jdk/commit/a7157eb0ef1b7190aabfb2689539801c6692bbae

For that particular key set, HashMap is faster, but I've found it really depends on the key set. However, this is mostly to demonstrate that emulation can have competitive performance with native `lookupswitch`. e.g. to get constant folding for constant inputs another filter has to be used, since C2 can not see through the HashMap lookups.

forax commented:
Ok, let restart from the beginning,
you have two strategy to de-sugar a switch,
- if what you do after the case do not mutate any variables, you can desugar each case to a method more or less like a lambda (it's not exactly like a lambda because there is no capture) and you have an indy in front that will call the right method handles
- you have a front end, with an indy that associate the object to an int and a backend which is tableswitch in the bytecode

The first strategy is an optimization, it will get you good performance by example if you compare a switch on a hirerachy on types and the equivalent method call. But you can not use that strategy for all switch, it's more an optimization.
The second strategy let you encode any switches.

The tests above are using the first strategy, which I think is not what we should implement by default given that it doesn't work will all cases. In the particular case of a switch on string, javac generates two switches, the front one and the back one, if we want to compare, we should implement the second strategy, so indy or the equivalent constant handle should take a String as parameter and return an int.

On the test themselves, for the hash, the Map should be directly bound, it will be more efficient, the asm version doesn't not appear in the graphics and there is a missing strategy that is using a MutableCallSite to switch from the a cascade of guards using the real values (store the String value even if it goes to `default`) and then switch to a lookup switch which i've found is the optimal strategy in real code (it works a lot like a bi-morphic cache but on string values instead of on classes).

JornVernee commented:
> you have two strategy to de-sugar a switch,
> 
> * if what you do after the case do not mutate any variables, you can desugar each case to a method more or less like a lambda (it's not exactly like a lambda because there is no capture) and you have an indy in front that will call the right method handles
> 
> * you have a front end, with an indy that associate the object to an int and a backend which is tableswitch in the bytecode
> 
> ...
>
> The tests above are using the first strategy

No, they are using the second strategy. The SwitchBootstraps patch I linked to replaces the front end `lookupswitch` of a String switch with an `invokedynamic` that computes an index for the back end jump table, which is still a `tableswitch` in the bytecode.

As John also described, a hypothetical lookupSwitch combinator can be emulated by using a `k -> [0, N)` projection that feeds into the tableSwitch combinator that is proposed by this PR. The point of the examples I linked to was to show several flavors of projection functions as an example of how this could be implemented, and to show that they have competitive performance with a native `lookupswitch` instruction (the 'legacy' case). i.e. the benchmarks show the difference between `lookupswitch` implemented in bytecode, and a `k -> [0, N)` projection function built by an `invokedynamic`. (sorry, I should have offered more explanation in the first place)

The combinator added by _this_ PR is not meant to replace any part of the String switch translation. For pattern switch the `tableSwitch` combinator _could_ be used to implement the front end `k -> [0, N)` projection, but it is not strictly required. Either way, that seems orthogonal to this PR.

JornVernee commented:
> About your benchmark, did you test with some strings going into "default", because it is usually in that case that you need a proper lookup switch,
another way to say it is that, your results are too good when you use a cascade of guardWithTest.

Yes, for the benchmarks I ran, the default case was just as likely as the other cases, so e.g. if there were 10 cases, there was a 1/11 chance the default case was hit. This might need tweaking to be more realistic but...

Note that the cascading guard with test actually works more like a binary search, where each guard tests against a pivot point in the search, and then decides to go either to the left or the right side of the tree. So, when looking up the default value we don't necessarily need to do a search over all the cases. Only for hash collisions does it fall back to a linear search over all the values with the same hash code.

This is also how C2 translates `lookupswitch` as far as I know (but maybe John can confirm or deny whether my reading of the C2 code is correct), so I'm not surprised to see that the if-tree approach is so close to the a native `lookupswitch` instruction in performance.

JornVernee commented:
I've just gotten back after 3 weeks of being sick, and would like to try moving this PR forward again.

Are there any remaining concerns with adding a tableSwitch combinator?

Reading back some of the discussion, I think the remaining point of contention is about adding a lookupSwitch combinator as well, which I think is a good idea (as a followup) in order to expose the lookupswitch bytecode as a combinator as well, but which doesn't seem like a blocker for this patch.

forax commented:
I hope you are well now.
You are right, adding a lookupswitch can be done later, i'm fine with the current state of this patch.

cl4es commented on src/java.base/share/classes/java/lang/invoke/MethodHandleImpl.java:
```diff
@@ -1952,6 +1959,142 @@ private static LambdaForm makeCollectorForm(MethodType basicType, Class<?> array
+        }
+    }
+
+    private static final Map<TableSwitchCacheKey, LambdaForm> TABLE_SWITCH_LAMBDA_FORM_CACHE = new ConcurrentHashMap<>();
```
This could be moved to `class TableSwitchCacheKey` to reduce eagerly initialization on bootstrap.

cl4es commented on src/java.base/share/classes/java/lang/invoke/LambdaForm.java:
```diff
@@ -1068,20 +1095,28 @@ boolean contains(Name name) {
         private @Stable MethodHandle resolvedHandle;
         @Stable MethodHandle invoker;
         private final MethodHandleImpl.Intrinsic intrinsicName;
+        private final Object intrinsicData;
```
This seems like the wrong place to add arbitrary data related only to intrinsics. Could this be moved either into `MethodHandleImpl$IntrinsicMethodHandle` or - perhaps less appealingly - rewrite the `MethodHandleImpl.Intrinsic` to a regular class which can then be instantiated with extra data? That `NamedFunction` holds a field of type `MethodHandleImpl.Intrinsic` instead of delegating to `resolvedHandle.intrinsicName()` seem like a pre-existing kludge that would be good to sort out why/if it's really necessary.

cl4es commented on src/java.base/share/classes/java/lang/invoke/InvokerBytecodeGenerator.java:
```diff
@@ -1393,6 +1399,58 @@ private static int popInsnOpcode(BasicType type) {
 
+    private Name emitTableSwitch(int pos, int numCases) {
+        Name args    = lambdaForm.names[pos];
+        Name invoker = lambdaForm.names[pos+1];
```
Add spaces around `+`

cl4es commented on src/java.base/share/classes/java/lang/invoke/InvokerBytecodeGenerator.java:
```diff
@@ -1393,6 +1399,58 @@ private static int popInsnOpcode(BasicType type) {
+
+        Class<?> returnType = result.function.resolvedHandle().type().returnType();
+        MethodType caseType = args.function.resolvedHandle().type()
+            .dropParameterTypes(0,1) // drop collector
```
Space after comma

cl4es commented on src/java.base/share/classes/java/lang/invoke/LambdaForm.java:
```diff
@@ -707,6 +708,32 @@ boolean isTryFinally(int pos) {
+        return tableSwitch.refersTo(MethodHandleImpl.class, "tableSwitch") &&
+                collectArgs.isInvokeBasic() &&
+                unboxResult.isInvokeBasic() &&
+                tableSwitch.lastUseIndex(collectArgs)  == 3 &&    // t_{n+1}:L=MethodHandleImpl.<invoker>(*, *, *, t_{n});
```
Extraneous space

cl4es commented on test/micro/org/openjdk/bench/java/lang/invoke/MethodHandlesTableSwitchRandom.java:
```diff
@@ -0,0 +1,149 @@
+        "25",
+        "50",
+        "100"
+    })
```
To reduce default runtime I think it'd be good to slim down the number of sizes we test here to 2 or 3 different ones.

cl4es commented on test/micro/org/openjdk/bench/java/lang/invoke/MethodHandlesTableSwitchOpaqueSingle.java:
```diff
@@ -0,0 +1,134 @@
+        "25",
+        "50",
+        "100"
+    })
```
To reduce default runtime I think it'd be good to slim down the number of sizes we test here to 2 or 3 different ones.

cl4es reviewed (changes requested):
Overall this looks great! 

I have a number of nits and a request to try and remodel so that `NamedFunction` isn't responsible for holding the arbitrary intrinsic data that you want to add here.

JornVernee commented:
Thanks for the review @cl4es 

I've addressed your review comments.

I've reduced the number of cases in the benchmark to 5, 10, and 25, which is the most interesting area to look at, and also removed the offset cases for the non-constant input benchmarks (proving the offset is constant folded only needs to be done once I think).

WRT `NamedFunction::intrinsicName`, I think you're right that we can also just delegate to `IntrinsicMethodHandle::intrinsicName`, and have it indicate the intrinsic. I've implemented that change. As a result, some `NamedFunction` constructor call sites no longer attach the intrinsic name to the `NamedFunction` itself, but re-wrap the `resolvedHandle` they use in an `IntrinsicMethodHandle` with the right intrinsic name. This leads to a nice code simplification from being able to remove all the `NamedFunction` constructor overloads. java/lang/invoke tests are still green, but I'll re-run Tier 1-3 as well to make sure that the difference in `resolvedHandle` being used for some `NamedFunctions` doesn't cause any other problems.

cl4es reviewed (approved):
Thanks for working through my comments, especially cleaning up how intrinsic data is handled which now looks much more like a natural fit!

JornVernee commented:
I've added a usage example to the javadoc in response to a review comment on the CSR.
